### ISSUE TITLE
Allow jsx fragments not to be aligned

### ISSUE THREAD

**Issue by vadzim** [yannickcr/eslint-plugin-react#2019]
This is a feature request.

Allow opening and closing tags for jsx fragments not to be on the same columns while require other jsx elements to be aligned. The purpose is to allow jsx fragments to act as brackets because `<>` and `</>` are small enough and could be used as brackets.

This can be achieved by handling a parameter for `react/jsx-closing-tag-location`, which disables aligning of fragments, but leaves aligning for components.

Example:

Let's there is a component with one child:
```javascript
const Component = ({
  ...props
}) => (
  <Child1>
    ...
  </Child1>
)
```
if now it's needed to add another child then it's needed to convert children to be a fragment:
```javascript
const Component = ({
  ...props
}) => (
  <>
    <Child1>         <-- extra change in git diff
      ...            <-- extra change in git diff
    </Child1>        <-- extra change in git diff
    <Child2>
      ...
    </Child2>
  </>
)
```
and so Child1 should indented to the right and this leads to not needed extra changes in git diff. This could be eliminated if it would be allowed for fragments to behave like brackets and to allow them to be unaligned like this
```javascript
const Component = ({
  ...props
}) => <>
  <Child1>         <-- marked as unchanged in git diff
    ...            <-- marked as unchanged in git diff
  </Child1>        <-- marked as unchanged in git diff
  <Child2>
    ...
  </Child2>
</>
```
There are other approaches, but they look not so good.
- Leaving `<>` on its own line, but not indented:
```javascript
const Component = ({
  ...props
}) =>
<>
  <Child1>
    ...
  </Child1>
  <Child2>
    ...
  </Child2>
</>
```
this also causes lint errors.
- turn off the linter rule `react/jsx-closing-tag-location`. This leads to too unaligned code.

**Comment by ljharb** [comment]
I don’t agree that fragment syntax should be treated the same as array brackets - the first example imo is the correct one.

It would add a lot of complexity to the rule to make alignment options be different for any arbitrary element, fragment syntax or not, and i don’t think it’s warranted.

**Comment by vadzim** [comment]
@ljharb I believe this question is of the same kind of using semicolons, tabs/spaces and so on. For me it is worth to have less noise in git diff. For another human it is worth to see jsx to be aligned.

Personally I will be happy with partially solution if `react/jsx-closing-tag-location` will be `true` by default but can take a parameter to disable it for fragments.

**Comment by ljharb** [comment]
I don’t think it’s the same - which indentation kind to alwsys use, or whether to always or never use semis, aren’t the same as “let’s treat this one jsx element type differently than all the others”.

**Comment by ljharb** [comment]
Fragment syntax is *nothing more than sugar* for React.Fragment, so it absolutely shouldn’t be treated differently there.

**Comment by vadzim** [comment]
I don't agree that it is nothing more than sugar.
React.Fragment is a very special type of a component. It looks like a component, but actually it is not a component. It is impossible to develop something that behaves like React.Fragment, it is a built-in feature of react. The only reason, why fragments can be represented with a pseudo-component React.Fragment is to provide compatibility with existent tools and to give some time for those tools to adopt to a shorthand syntax. Even possibility of using React.Fragment for key prop is just a trade-off for syntax simplicity and usability.

**Comment by ljharb** [comment]
That’s an argument for both fragment syntax *and* React.Fragment to be treated differently than other elements (although that means the *component* is different, but in no way is the *element* different), but not an argument to treat fragment syntax differently than React.Fragment.

**Comment by ljharb** [comment]
Also, you’re wrong; React.Fragment can be fully polyfilled in React 16.0 or 16.1, and even partially polyfilled in any React.

**Comment by vadzim** [comment]
I don't think my arguments is about shorthand and React.Fragment. Eslint is about syntax and looking, and `<React.Fragment...` looks like a component, but `<>` does not.

Anyway it can be just a bug of my eyes, but I believe there exists bunch of people with the same feature in their vision system.

React 16.0 supports arrays AFAIK, it was already prepared to support fragments.
And I really cannot imagine the way fragments could be implemented in React 15. I had tried to do this several times)

**Comment by ljharb** [comment]
You’d do it by wrapping them in a div, basically.

I understand the aesthetic argument - but i think fragment syntax looks much more like an element than like an opening array bracket.

**Comment by ljharb** [comment]
Then you could do it with mutating the DOM directly and using document fragments ¯\_(ツ)_/¯ there’s lots of ways, but it’s not really the point.
